### PR TITLE
Add function to compute incidence matrix

### PR DESCRIPTION
**Pull request by leethargo**
It is now implemented only for directed graphs, which I find most useful, but I could also do it for directed graphs.

Also, I'm a Julia newbie, so I'm happy about feedback.

### CODE REVIEW

**Comment by jpfairbanks** [comment]
This looks right. My only feedback are on the docs and performance. 

The docs say matrix indexed by [v,e]. It is defined that way but the matrix you get is indexed by [v,i] where i is an Int representing the order of traversal of the graph. The edge iterator might change that order in the future and the docs should make it clear that you can't index the SparseMatrixCSC by a vertex, edge pair.

The performance issue is probably not important, but you can avoid the `push!` calls by iterating over the graph with this construction.

```julia
for (i, e) in enumerate(edges(g))
    #access indices using i, 2i, and 2i+1
end
```

I think that would also make the structure of the matrix more explicit from the code.

File: src/spectral.jl [hunk 1]
```diff
@@ -133,3 +133,27 @@ function CombinatorialAdjacency(g::Graph)
...
+"""
+function incidence_matrix(g::DiGraph, T::DataType=Int)
+    n_v = nv(g)
+    n_e = ne(g)
+    nz = 2 * ne(g)
+    colpt = collect(1:2:(2*n_e + 1)) # every col has 2 entries
+    rowval = sizehint!(Vector{Int}(), nz)
+    nzval = sizehint!(Vector{T}(), nz)
+    p_one, m_one = one(T), -one(T)
+    for e in edges(g)
```
Comment (sbromberger) [inline src/spectral.jl]: In general, iteration over edges is suboptimal as far as performance is concerned. Is it possible to use the adjacency lists (`fadj` and `badj`) directly?

File: test/spectral.jl [hunk 2]
```diff
@@ -67,3 +67,9 @@ end
     @test_approx_eq_eps(eigs(lapl, which=:LR)[1][1], 3.902, 0.001)
     println("*** Finished GraphMatrices tests")
 end
+
+# testing incidence_matrix
+@test size(incidence_matrix(g4)) == (5,4)
+@test incidence_matrix(g4)[1,1] == -1
+@test incidence_matrix(g4)[2,1] == 1
+@test incidence_matrix(g4)[3,1] == 0
```
Comment (sbromberger) [inline test/spectral.jl]: :+1: for reusing existing graphs. :)

**Comment by leethargo** [comment]
Thanks for all your feedback!

I agree about the documentation. It shouldn't talk about edges, since these are typically pairs of vertices in LightGraphs.

Is there a proper ordering of the edges? Otherwise it might be useful to sort them lexicographically.

I don't think I understand the part about the `enumerate`. Dou you mean that I could write instead:
```
rowval[2i]     = src(e)
rowval[2i + 1] = dst(e)
```
and that would be more efficient than the `push!`?

I guess I can also use the `fadj` directly, I'll look into that.

For undirected graphs, the sparsity pattern is the same, but it's `1` instead of `-1`. I'm not sure what to do about self-loops.

Another question: Should the function be moved somewhere else, since it doesn't relate to `spectral` directly?

And: should I be concerned that the Travis CI build failed?

**Comment by leethargo** [comment]
OK, I've made some of the changes. I guess that when I iterate over `fadj`, I can no longer use the `enumerate` trick, right?

**Comment by jpfairbanks** [comment]
> Is there a proper ordering of the edges? Otherwise it might be useful to sort them lexicographically.

They currently are sorted that way but we don't want client code to depend on that for correctness because we might change it. That is my preference although I don't see it documented anywhere.

> I don't think I understand the part about the enumerate. Do you mean that I could write instead:

```julia
for (i,e) in enumerate(edges(g))
   rowval[2i]     = src(e)
   rowval[2i + 1] = dst(e)
end
nzval = repmat([1,-1], ne(g))
```
> and that would be more efficient than the push!?

`push!` has to check every time that allocated memory is large enough for the new data and then reallocate if necessary. In this case it is never necessary since `sizehint!` is called with the correct size of the final array. So you can avoid the checks and just write the value to the memory locations `rowval[2i]` and `rowval[2i+1]`.


> I guess I can also use the fadj directly, I'll look into that.

I think seth wants to avoid people using the internal structures because then we have less freedom in changing it. I would say only use fadj directly if you have benchmarks that show you get substantially better performance.

> For undirected graphs, the sparsity pattern is the same, but it's 1 instead of -1. I'm not sure what to do about self-loops.

I have always seen undirected graph incidence matrices with arbitrary direction on the edges rather than an all positive matrix. Don't you need to put signs on in order to preserve the definition of the Laplacian as the incidence matrix times its transpose?

> Another question: Should the function be moved somewhere else, since it doesn't relate to spectral directly?

I am inclined to have everything related to matrix representations of graphs in spectral.jl. But my preference is weak.


> And: should I be concerned that the Travis CI build failed?

eh, probably. although it is on nightly so it might not be your fault.

**Comment by jpfairbanks** [comment]
I just saw seth's line comment about fadjlist, I guess we are out of sync on that issue. :laughing: 
You can still remove the `push!` by counting which edge you are on yourself. 
```julia
i = 0
for v in vertices(g)
    for u in neighbors(g,v)
         i+=1
    end
end
```

**Comment by jpfairbanks** [comment]
Diestel's Graph theory book uses your definition so we can use that def.

The directedness check can be lifted out of the loop because it doesn't change.
Either define th

**Comment by sbromberger** [comment]
> I just saw seth's line comment about fadjlist, I guess we are out of sync on that issue. :laughing: 

So, yes - we don't want folks to use `fadjlist` and `badjlist` - that's why we expose forward and backward adjacencies via the `fadj` and `badj` functions. Those are acceptable APIs. :)

**Comment by sbromberger** [comment]
> And: should I be concerned that the Travis CI build failed?

I don't think so, since (unless I'm seeing something else) it looks like GZip.jl is breaking the Net file format.

**Comment by leethargo** [comment]
OK, I'll avoid the `push!` using the counter. The check `isdirected` was already outside the loop.

**Comment by leethargo** [comment]
Hm, so apparently the version for directed graphs *doesn't* work as expected, because every edge occurs twice in the adjacency list. I'm not really sure how to fix this right now without resolving to the edge iterator, as before.

**Comment by jpfairbanks** [comment]
> because every edge occurs twice in the adjacency list

Can you write a unit test that shows this?

**Comment by sbromberger** [comment]
> Hm, so apparently the version for directed graphs doesn't work as expected, because every edge occurs twice in the adjacency list.

I don't think that's correct. The forward adjacency for a given vertex will only contain those destination vertices that make up an edge; similarly, the backward adjacency will only contain source vertices for an edge. That is, you should probably use `+1` for one of them and `-1` for the other.

File: src/spectral.jl [hunk 3]
```diff
@@ -133,3 +133,33 @@ function CombinatorialAdjacency(g::Graph)
...
+
+    # every col has the same 2 entries
+    u_coef = is_directed(g)? -one(T) : one(T)
+    v_coef = one(T)
+    colpt = collect(1:2:(nz + 1))
+    nzval = repmat([u_coef, v_coef], n_e)
+
+    # iterate over edges for row indices
+    rowval = zeros(Int, nz)
+    for (i,e) in enumerate(edges(g))
```
Comment (sbromberger) [inline src/spectral.jl]: so you're still iterating over edges here - this is likely to be give poor performance. Can you try `@time` with medium-sized (1e6,1e7) graphs?

**Comment by leethargo** [comment]
So, I don't believe it's a bug, just a misunderstanding of what's stored in `fadjlist` for directed graph. Consider this example:
```
julia> g = Graph(2)
{2, 0} undirected graph

julia> add_edge!(g, 1, 2)
edge 1 - 2

julia> ne(g)
1

julia> for u in vertices(g)
       for v in out_neighbors(g, u)
       println("$u, $v")
       end
       end
1, 2
2, 1
```

The graph has only one edge (1-2), but when I loop over the `out_neighbors` (or `fadj`) then I also get the reversed edge (2-1). That's why I got out of bounds problems in my code.

I'd need a way to tell whether I had already seen an edge before. Maybe I can check for `u < v` and only then add the entries?

**Comment by leethargo** [comment]
The version with `fadj` is 3-5 times faster on [this benchmark](https://gist.github.com/leethargo/c9db87d1351b95d1918d).

But the `u < v` check is not valid for directed graphs, so I'll two methods now.

**Comment by sbromberger** [comment]
@Leethargo - the example graph you posted is an undirected graph, which means that you are seeing both 1 and 2 as source vertices and therefore they will both have forward adjacencies to each other. 

A directed graph with the same edge added will have one fadj from 1 to 2 and one badj from 2 to 1.

**Comment by leethargo** [comment]
@sbromberger : Exactly, I meant to write undirected and fixed my comment. Even for undirected graphs, the incidence matrix should only have `ne(g)` many columns. Therefore, I've skipped all edges that are not ordered.

From my point of view, the behaviour is now correct.

There is still a lot of code shared between the two methods, but I thought that factoring these parts out will not make it more readable.

**Comment by sbromberger** [comment]
We have a function `is_ordered` that should be available for use. (If it's not exported, it should be. ). That's probably the right way of doing it if you're not doing it that way already.

**Comment by sbromberger** [comment]
PS: as you've seen, undirected graphs don't have a badjlist. They use a single data structure to hold all adjacency info. This means that they're as memory efficient, per edge, as directed graphs. `badj()` is just a pointer to `fadj()` for undirected graphs.

**Comment by leethargo** [comment]
OK, it now uses `is_ordered`.

File: src/spectral.jl [hunk 4]
```diff
@@ -133,3 +133,62 @@ function CombinatorialAdjacency(g::Graph)
...
+    colpt = collect(1:2:(nz + 1))
+    nzval = repmat([u_coef, v_coef], n_e)
+
+    # iterate over edges for row indices
+    rowval = zeros(Int, nz)
+    i = 1
+    for u in vertices(g)
+        for v in out_neighbors(g, u)
+            if is_ordered(Edge(u, v)) # add every edge only once
+                rowval[2*i - 1] = u
```
Comment (sbromberger) [inline src/spectral.jl]: Question: why not combine these two methods and stick this check in the combined method, sort of like this?

Also, I'm taking back what I said about using `is_ordered` here. It's much less efficient to create an Edge and use `is_ordered()` than it is to just check `u < v`.

```
isdir = is_directed(g)
```
then, in this loop,
```
   isdir && u < v && ...
```

**Comment by leethargo** [comment]
I like that better. It should be `isdir || u < v` though, I believe.

**Comment by sbromberger** [comment]
> I like that better. It should be `isdir || u < v` though, I believe.

Yeah, I hadn't given it too much thought for correctness but I wanted to get the point across that you can likely combine these things with minimal effort. Putting the `isdir` check in there is not going to slow things down a whole lot since it will be inlined and probably optimized out.

When you're ready to go, would you mind squashing the commits and we'll have a final runthrough?

Thanks for all your work on this. PS: spectral is probably the best place for this sort of stuff since it's "matrixy" (not that there's anything wrong with that but I'm a structural guy; @jpfairbanks is our resident spectral guru).

**Comment by leethargo** [comment]
Sure, I pushed the squashed commit to the same branch, hopefully that doesn't break the pull request.

PS: The incidence matrix is actually related to the Laplacian, but I want to use in the context of network flows.

**Comment by sbromberger** [comment]
PS: hold off a minute - we'll want to retest - I think I found the error that's causing travis to fail (see #282).

**Comment by sbromberger** [comment]
restarted the test. this should pass.

**Comment by sbromberger** [comment]
All good! Thanks again.

**Comment by jpfairbanks** [comment]
Thanks for going through this and getting it right! 

I think that spectral.jl is a good place for any graphs and matrices stuff. I'm going with the nomenclature of Spectral Graph Theory which is using matrix properties to prove things about graphs rather than a specifically eigenvalues and eigenvectors definition.

If spectral.jl really fills up in the future we could revisit that decision.

I'm glad to have someone working on flows involved. I think that is an area that could use some improvement.

**Comment by leethargo** [comment]
> I'm glad to have someone working on flows involved. I think that is an area that could use some improvement.
I was planning to use JuMP for that, but we'll see.

**Pull request merged**